server/src/main/java/org/elasticsearch/common/settings/ConsistentSecureSettingsValidatorService.java

```diff
@@ -0,0 +1,257 @@
+            return key.getEncoded();
+        } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+            throw new RuntimeException("\"PBKDF2WithHmacSHA512\" algorithm is required for consistent secure settings", e);
+        }
+    }
+
+    private static String computePublicHash(char[] localHash, String salt) {
+        final byte[] saltedHashBytes = computeSaltedHash(localHash, salt.getBytes(StandardCharsets.UTF_8));
+        final String base64SaltedHashString = new String(Base64.getEncoder().encode(saltedHashBytes), StandardCharsets.UTF_8);
+        return salt + ":" + base64SaltedHashString;
+    }
+
+    private boolean checkPublicHashAgainstLocalValue(String publishedSettingName, String publishedHashWithSalt) {
+        logger.trace("published hash for secure setting [{}] is [{}]", publishedSettingName, publishedHashWithSalt);
+        if (publishedHashWithSalt == null || false == publishedHashWithSalt.contains(":")) {
+            logger.trace("published hash [{}] for secure setting [{}] is invalid", publishedHashWithSalt, publishedSettingName);
+            return false;
+        }
+        final String[] parts = publishedHashWithSalt.split(":");
+        if (parts == null || parts.length != 2) {
+            logger.trace("published hash [{}] for secure setting [{}] is invalid", publishedHashWithSalt, publishedSettingName);
+            return false;
+        }
+        final char[] localHash = localHashesOfConsistentSettings.get(publishedSettingName);
+        if (localHash == null) {
+            logger.trace("secure setting [{}] is missing on the local node", publishedSettingName);
+            return false;
+        }
+        final String publishedSalt = parts[0];
+        final String localHashWithSalt = computePublicHash(localHash, publishedSalt);
+        logger.trace("local hash for secure setting [{}] is [{}]", publishedSettingName, localHashWithSalt);
+        return publishedHashWithSalt.equals(localHashWithSalt);
+    }
+
+    Map<String, String> publicHashesOfConsistentSettings() {
+        return publicHashesOfConsistentSettings;
+    }
+
+    @Override
+    public void applyClusterState(ClusterChangedEvent event) {
+        boolean allConsistent = true;
+        final Map<String, String> publishedHashesOfConsistentSettings = event.state().metaData().hashesOfConsistentSettings();
+        if (event.localNodeMaster()) {
+            logger.debug("local node is elected master, no hashes consistency check is required");
+        } else if (publishedHashesOfConsistentSettings.equals(publicHashesOfConsistentSettings)) {
+            logger.debug("hashes of secure settings are identical to the master's");
+        } else {
+            for (Map.Entry<String, String> publishedSettingAndHash : publishedHashesOfConsistentSettings.entrySet()) {
+                if (false == checkPublicHashAgainstLocalValue(publishedSettingAndHash.getKey(), publishedSettingAndHash.getValue())) {
```

Comment:
I think we need to switch threads to compute the hashes here (when we need the computation) since this runs on the cluster state applier thread and we shouldn't block that thread. At this point it is not an issue since we would have a single setting but its kind of a ticking time bomb as more consistent settings are added